File: Week_04/kotlin/src/_367_ValidPerfectSquare.kt
fun isPerfectSquare(num: Int): Boolean {
    if (num < 1) return false
    if (num == 1) return true
    var l = 1
    var h = num
    while (l < h) {
        val m = l + ((h - l) ushr 1)
        val r = 1L * m * m
        if (r == num.toLong()) return true
        if (r < num) l = m + 1 else h = m
    }
    return false
}

fun isPerfectSquare1(num: Int): Boolean {
    if (num < 1) return false
    var r = num.toLong()
    while (r * r > num) r = (r + num / r) ushr 1
    return r * r == num.toLong()
}

fun main() {
    println("a")
}

File: Week_04/cpp/src/_367_ValidPerfectSquare.cpp
class UsingNewtonMethod
{
public:
    bool isPerfectSquare(int num)
    {
        if (num < 0)
            return false;
        long long x = num;
        while (x * x > num)
            x = (x + num / x) >> 1;
        return x * x == num;
    }
};
class UsingBinarySearch
{
public:
    bool isPerfectSquare(int num)
    {
        if (num < 2)
            return true;
        int l = 0, r = num;
        while (l <= r)
        {
            long m = l + ((r - l) >> 1);
            if (m * m == num)
                return true;
            if (m * m > num)
                r = m - 1;
            else
                l = m + 1;
        }
        return false;
    }
};